cpp/unreal_plugins/SimulationController/Source/SimulationController/NavMesh.cpp

```diff
@@ -83,3 +90,72 @@ std::vector<uint8_t> NavMesh::getRandomPoints(int num_points)
     }
     return Std::reinterpretAs<uint8_t>(points);
 }
+
+std::vector<uint8_t> NavMesh::getReachablePoints(const std::vector<std::vector<float>>& start_points)
+{
+    std::vector<double> end_points;
+    for (auto& start_point : start_points) {
+
+        SP_ASSERT(start_point.size()==3);
+
+        FVector start_point_fvector{start_point.at(0), start_point.at(1), start_point.at(2)};
+        FNavLocation end_location;
+
+        bool found = recast_nav_mesh_->GetRandomReachablePointInRadius(
+            start_point_fvector, Config::get<float>("SIMULATION_CONTROLLER.NAVMESH.TRAJECTORY_SAMPLING_SEARCH_RADIUS"), end_location);
+        SP_ASSERT(found);
+
+        end_points.push_back(end_location.Location.X);
+        end_points.push_back(end_location.Location.Y);
+        end_points.push_back(end_location.Location.Z);
+    }
+    return Std::reinterpretAs<uint8_t>(end_points);
+}
+
+std::vector<std::vector<uint8_t>> NavMesh::getTrajectories(const std::vector<std::vector<float>>& start_points, const std::vector<std::vector<float>>& end_points)
+{
+    SP_ASSERT(start_points.size() == end_points.size());
+
+    std::vector<std::vector<uint8_t>> trajectories;
+
+    for (int i = 0; i < start_points.size(); ++i) {
+        SP_ASSERT(start_points.at(i).size() == 3);
+        SP_ASSERT(end_points.at(i).size() == 3);
+
+        FVector start_point_fvector{start_points.at(i).at(0), start_points.at(i).at(1), start_points.at(i).at(2)};
+        FVector end_point_fvector{end_points.at(i).at(0), end_points.at(i).at(1), end_points.at(i).at(2)};
+
+        // Update navigation query with the start and end location
+        FPathFindingQuery nav_query = FPathFindingQuery(world_, *recast_nav_mesh_, start_point_fvector, end_point_fvector);
+
+        // Generate a collision-free path between the start location and the end location
+        FPathFindingResult path = navigation_system_v1_->FindPathSync(nav_query, EPathFindingMode::Type::Regular);
+
+        // Ensure that path generation process was successful and that the generated path is valid
+        SP_ASSERT(path.IsSuccessful());
+        SP_ASSERT(path.Path.IsValid());
+
+        // Update trajectory with the waypoints
+        TArray<FNavPathPoint> path_points = path.Path->GetPathPoints();
+        SP_ASSERT(path_points.Num() > 1);
+
+        std::vector<double> trajectory;
+        for (auto& path_point : path_points) {
+            trajectory.push_back(path_point.Location.X);
+            trajectory.push_back(path_point.Location.Y);
+            trajectory.push_back(path_point.Location.Z);
+        }
+
+        trajectories.push_back(Std::reinterpretAs<uint8_t>(trajectory));
+
+        // Debug drawings
```

Comment:
Remove this debug drawing functionality. It's too application-specific for this general-purpose scene querying functionality. We're not really using it for anything currently. And I'd rather move this debug draw functionality into a general-purpose `DebugDraw` interface (similar to this `NavMesh` interface), where users could call from their Python code if they wanted to.

Reply:
Updated this.